### ISSUE TITLE
'Wrong transactional state (no active transaction, wrong type of transaction, or transaction already in progress)

### ISSUE THREAD

**Issue by mr-grey** [realm/realm-swift#7912]
### How frequently does the bug occur?

Sometimes

### Description

used extension 

**all write operations make at queue** 

```
   static var operationQueue: OperationQueue {
        get {
          queue.qualityOfService = .userInteractive
          queue.maxConcurrentOperationCount = 1
          return queue
        }
    }
    static let serialQueue = DispatchQueue.init(label: "serial-queue", qos: DispatchQoS.userInitiated) // ....

   // open realm 
    serialQueue.async {
            do {
                RealmManager.sharedRealm = try Realm(configuration: .defaultConfiguration, queue: serialQueue)
            } 
   }

```

```
extension Realm {
    public func safeWrite(_ block: (() throws -> Void)) throws {
        if isInWriteTransaction {
            try block()
        } else {
            try write(block)
        }
    }
}
```

> crash at line   try block()

that means  realm isInWriteTransaction == true 
how it possible 

Crash happened some time 

screenshots 

![Screenshot 2022-08-16 at 11 28 38](https://user-images.githubusercontent.com/699148/184834489-74643920-f0f4-4d39-8319-3513e0018810.png)
![Screenshot 2022-08-16 at 11 25 57](https://user-images.githubusercontent.com/699148/184834556-c419e342-793f-442b-8433-4704091e5676.png)
![Screenshot 2022-08-16 at 11 25 50](https://user-images.githubusercontent.com/699148/184834583-872783a8-e23c-4c84-9026-c22af9b0dca6.png)
![Screenshot 2022-08-16 at 11 26 11](https://user-images.githubusercontent.com/699148/184834619-d2141e34-cb81-4bc0-9941-e4cf046743cd.png)

![Screenshot 2022-08-16 at 11 25 57](https://user-images.githubusercontent.com/699148/184834747-fef646dd-20dc-4797-addb-416f8d639bb0.png)

### Stacktrace & log output

```shell
*** Terminating app due to uncaught exception 'RLMException', reason: 'Wrong transactional state (no active transaction, wrong type of transaction, or transaction already in progress)
Exception backtrace:
0   Realm                               0x000000010ac8cc08 _ZN5realm4util6detail26ExceptionWithBacktraceBaseC2Ev + 48
1   Realm                               0x000000010ac8cb6c _ZN5realm4util22ExceptionWithBacktraceISt9exceptionEC2IJEEEDpOT_ + 68
2   Realm                               0x000000010ac8cb00 _ZN5realm10LogicErrorC2ENS0_9ErrorKindE + 60
3   Realm                               0x000000010ac8ca20 _ZN5realm10LogicErrorC1ENS0_9ErrorKindE + 36
4   Realm                               0x000000010b0309d4 _ZN5realm4Node16do_copy_on_writeEm + 400
5   Realm                               0x000000010af96580 _ZN5realm5Array3setEmx + 124
6   Realm                               0x000000010b03092c _ZN5realm4Node16do_copy_on_writeEm + 232
7   Realm                               0x000000010af96580 _ZN5realm5Array3setEmx + 124
8   Realm                               0x000000010ac743a4 _ZN5realm5Array16update_child_refEmm + 40
9   Realm                               0x000000010b03092c _ZN5realm4Node16do_copy_on_writeEm + 232
10  Realm                               0x000000010af96580 _ZN5realm5Array3setEmx + 124
11  Realm                               0x000000010afc3464 _ZN5realm16ClusterNodeInner20update_ref_in_parentENS_6ObjKeyEm + 88
12  Realm                               0x000000010b031250 _ZN5realm3Obj3setINS_10StringDataEEERS0_NS_6ColKeyET_b + 864
13  Realm                               0x000000010ac8eb48 _ZZN12_GLOBAL__N_114setValueOrNullIN5realm10StringDataEEEvP13RLMObjectBaseNS1_6ColKeyEP11objc_objectENKUlvE_clEv + 136
14  Realm                               0x000000010ac8ea34 _ZL17RLMTranslateErrorIZN12_GLOBAL__N_114setValueOrNullIN5realm10StringDataEEEvP13RLMObjectBaseNS2_6ColKeyEP11objc_objectEUlvE_EDaOT_ + 24
15  Realm                               0x000000010ac8ea10 _ZN12_GLOBAL__N_114setValueOrNullIN5realm10StringDataEEEvP13RLMObjectBaseNS1_6ColKeyEP11objc_object + 88
16  Realm                               0x000000010ac8e9ac _ZN12_GLOBAL__N_18setValueEP13RLMObjectBaseN5realm6ColKeyEP8NSString + 48
17  Realm                               0x000000010ac679c4 _ZN12_GLOBAL__N_111kvoSetValueIU8__strongP8NSStringS3_EEvP13RLMObjectBasemT_ + 136
18  Realm                               0x000000010ac678fc RLMSetSwiftPropertyString + 72
19  RealmSwift                          0x000000010bf13c7c $sSS10RealmSwiftE15_rlmSetPropertyyySo13RLMObjectBaseC_s6UInt16VSStFZ + 124
20  RealmSwift                          0x000000010bf13d2c $sSS10RealmSwift12_PersistableA2aBP15_rlmSetPropertyyySo13RLMObjectBaseC_s6UInt16VxtFZTW + 24
21  RealmSwift                          0x000000010bfa58b0 $s10RealmSwift9PersistedV3set_5valueySo13RLMObjectBaseC_xtF + 2076
22  RealmSwift                          0x000000010bfa502c $s10RealmSwift9PersistedV18_enclosingInstance7wrapped7storagexqd___s24ReferenceWritableKeyPathCyqd__xGAHyqd__ACyxGGtcSo13RLMObjectBaseCRbd__luisZ + 228
23  Me                                  0x0000000101021a10 $s2Me9RealmUserC9firstNameSSvs + 156
24  Me                                  0x000000010103d01c $s2Me9RealmUserC7setJSON33_28A8C472E5AD302037F6F876A8769A88LL_7toModely06SwiftyE00E0V_ACtFZ + 1256
25  Me                                  0x000000010103ca40 $s2Me9RealmUserC6upsert4jsony10SwiftyJSON0G0V_tFZyycfU_yyXEfU_ + 64
26  Me                                  0x00000001014ba8b8 $s10RealmSwift0A0V2MeE9safeWriteyyyyKXEKF + 112
```

### Can you reproduce the bug?

Yes, sometimes

### Reproduction Steps

multiple transactions at background 

### Version

  pod 'RealmSwift', '~>10.28.2'

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

iOS 12, 14, 15

### Build environment

Xcode version: ... Version 13.3.1 (13E500a)

**Comment by leemaguire** [comment]
@mr-grey having the serial queue being dispatched inside another operation queue seems like it might be the culprit to the issue. Could you try removing `operationQueue`? Also Realms are cached so I would not recommend creating a singleton manager class to access a Realm. You should create transient instances to the Realm when you need to access it.

**Comment by mr-grey** [comment]
> 

Hi,  1. operationQueue was added for prevent inTransaction state . because multiple different transactions in threads make crash .
2. This issue 'Wrong transactional state..' occurs only for old devices , possible it unable to write to file or some like this 
   am modified a extension for it wait for InTransaction become false. 

```
public func safeWrite(_ block: (() throws -> Void)) throws {

        //time in microseconds instead

        while isInWriteTransaction {

            usleep(10000)

        }

        try write(block)
    } 
```

3. 'Realms are cached so I would not recommend creating a singleton manager class to access a Realm. ' - one Realm used for all write operations.  Also how about memory consumption, on older devices it a problem , in first time we have used singe Realm  for single get \ write data operations. but it provides multiple crashed via memory issues .
